utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala

```diff
@@ -136,59 +136,71 @@ object SchemaPathValidator {
     }
   }
 
+  @tailrec
   private def validateSchemaPathArray(schema: StructType,
                                       path: Array[String],
                                       parentOnly: Boolean = false,
                                       fullPathNew: Boolean = false,
                                       parentPath: String = ""): Seq[ValidationIssue] = {
-    // scalastyle:off  return Covered in Issue #730
     if (path.isEmpty) {
-      return Nil
-    }
-
-    val currentField = path(0)
-    val fullPath = s"$parentPath${path.mkString(".")}"
-
-    if (parentOnly && path.length == 1) {
-      if (fullPathNew) {
-        val exactMatch = schema.fields.find(field => field.name == currentField)
-        if (exactMatch.nonEmpty) {
-          return Seq(ValidationError(s"Column '$parentPath$currentField' already exists so it cannot be used as an output column '$fullPath'."))
-        }
-        val caseInsensitiveMatch = schema.fields.find(field => field.name.compareToIgnoreCase(currentField) == 0)
-        if (caseInsensitiveMatch.nonEmpty) {
-          return Seq(ValidationError(s"Case insensitive variant of a cloumn '$parentPath$currentField' already exists so it cannot be used as an output column '$fullPath'."))
+      Nil
+    } else {
+      val currentField = path(0)
+      val fullPath = s"$parentPath${path.mkString(".")}"
+      if (parentOnly && path.length == 1) {
+        if (fullPathNew) {
+          handleParentMatch(schema, parentPath, currentField, fullPath)
+        } else {
+          Nil
         }
-      }
-      return Nil
-    }
-
-    val exactMatch = schema.fields.find(field => field.name == currentField)
-    val failures = exactMatch match {
-      case Some(field) =>
-        val dataType = getUnderlyingType(field.dataType)
-        dataType match {
-          case st: StructType =>
-            validateSchemaPathArray(st, path.drop(1), parentOnly, fullPathNew, s"$parentPath${field.name}.")
-          case _ =>
-            if (path.length > 1) {
-              Seq(ValidationError(s"Column '$parentPath$currentField' is a primitive type and can't contain child fields '$fullPath'."))
-            } else {
-              Nil
+      } else {
+        val matched = checkMatchType(schema, currentField)
```

Comment:
I would think about even extracting this huge else. If you decide to keep it, keep it. I think there are arguments for both.

Reply:
I tried to think of a way to simplify and extract, but I could not find one without overcomplicating or extracting to a method with a lot of parameters